packages/dds/tree/src/simple-tree/core/treeNodeKernel.ts

```diff
@@ -354,6 +368,249 @@ const kernelEvents = ["childrenChangedAfterBatch", "subtreeChangedAfterBatch"] a
+		if (pauseTreeEventsStack) {
+			for (const fieldKey of changedFields) {
+				this.childrenChangedBuffer.add(fieldKey);
+			}
+		} else {
+			for (const listener of this.childrenChangedListeners) {
+				listener({ changedFields });
+			}
+		}
+	}
+
+	private handleSubtreeChangedAfterBatch(): void {
+		if (pauseTreeEventsStack) {
+			this.subTreeChangedBuffer = true;
+		} else {
+			for (const listener of this.subTreeChangedListeners) {
+				listener();
+			}
+		}
+	}
+
+	public on(
+		eventName: "childrenChangedAfterBatch",
+		listener: (arg: {
+			changedFields: ReadonlySet<FieldKey>;
+		}) => void,
+	): () => void;
+	public on(eventName: "subtreeChangedAfterBatch", listener: () => void): () => void;
+	public on(
+		eventName: keyof KernelEvents,
+		listener: (arg: {
+			changedFields: ReadonlySet<FieldKey>;
+		}) => void | (() => void),
+	): () => void {
+		this.assertNotDisposed();
+		switch (eventName) {
+			case "childrenChangedAfterBatch":
+				this.childrenChangedListeners.add(
+					listener as (arg: {
+						changedFields: ReadonlySet<FieldKey>;
+					}) => void,
+				);
+				return () => this.off("childrenChangedAfterBatch", listener);
+			case "subtreeChangedAfterBatch":
+				this.subTreeChangedListeners.add(listener as () => void);
+				return () => this.off("subtreeChangedAfterBatch", listener as () => void);
+			default:
+				unreachableCase(eventName);
+		}
+	}
+
+	public off(
+		eventName: "childrenChangedAfterBatch",
+		listener: (arg: {
+			changedFields: ReadonlySet<FieldKey>;
+		}) => void,
+	): void;
+	public off(eventName: "subtreeChangedAfterBatch", listener: () => void): void;
+	public off(
+		eventName: keyof KernelEvents,
+		listener: (arg: {
+			changedFields: ReadonlySet<FieldKey>;
+		}) => void | (() => void),
+	): void {
+		this.assertNotDisposed();
+		switch (eventName) {
+			case "childrenChangedAfterBatch":
+				this.childrenChangedListeners.delete(
+					listener as (arg: {
+						changedFields: ReadonlySet<FieldKey>;
+					}) => void,
+				);
+				break;
+			case "subtreeChangedAfterBatch":
+				this.subTreeChangedListeners.delete(listener as () => void);
+				break;
+			default:
+				unreachableCase(eventName);
+		}
+	}
+
+	/**
+	 * Flushes any events buffered due to {@link pauseTreeEvents}.
+	 */
+	public flush(): void {
+		this.assertNotDisposed();
+
+		if (this.childrenChangedBuffer.size > 0) {
+			for (const listener of this.childrenChangedListeners) {
+				listener({ changedFields: this.childrenChangedBuffer });
+			}
+			this.childrenChangedBuffer.clear();
+		}
+
+		if (this.subTreeChangedBuffer) {
+			for (const listener of this.subTreeChangedListeners) {
+				listener();
+			}
+			this.subTreeChangedBuffer = false;
+		}
+	}
+
+	private assertNotDisposed(): void {
+		assert(!this.disposed, "Event handler disposed.");
+	}
+
+	public dispose(): void {
+		if (this.disposed) {
+			return;
+		}
+
+		this.disposeOnFlushListener();
```

Comment:
Should childrenChangedListeners and subTreeChangedListeners be cleared after this dispose call?